### PR TITLE
[export] Fix torch.export() issues for server use cases.

### PR DESCRIPTION
**Pull request by zhxchen17**
Test Plan: In D48788843

Differential Revision: D48811793

### CODE REVIEW

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -433,24 +484,37 @@ def to_str_dict(sig_component: Dict[Any, Any]):
...
         [ModuleCallEntry(fqn, sig) for fqn, sig in module_call_signatures.items()],
         (args, {}),
     )
 
-    exported_program = exported_program._transform(
-        _AddRuntimeAssertionsForInlineConstraintsPass(range_constraints, equality_constraints)
-    )
-    if len(preserve_module_call_signature) > 0:
-        exported_program = exported_program._transform(CollectTracepointsPass(module_call_signatures))
+    if len(range_constraints) > 0 or len(equality_constraints) > 0:
```
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:513]: Do you know if we put the unbacked symint based ranges into above two lists?

File: torch/_export/passes/collect_tracepoints_pass.py [hunk 2]
```diff
@@ -18,6 +20,15 @@ def __init__(self, specs) -> None:
         self.specs = specs
 
     def call(self, gm):
+        def get_arg_spec(arg):
+            if isinstance(arg, torch.fx.Node):
+                if isinstance(arg.meta.get('val'), torch.Tensor):
+                    return ArgumentSpec(kind=ArgumentKind.Tensor, value=arg.name)
+                else:
+                    raise RuntimeError("NYI")
```
Comment (tugsbayasgalan) [inline torch/_export/passes/collect_tracepoints_pass.py:28]: NIT: Could you add a more descriptive error message?

File: torch/_export/__init__.py [hunk 3]
```diff
@@ -210,6 +210,36 @@ def _safe_to_skip_dynamo(gm: torch.fx.GraphModule):
...
+        for node in gm.graph.nodes:
+            if node.op in ["placeholder", "output"]:
+                continue
+            if nn_module_stack := node.meta.get("nn_module_stack", {}):
+                path, ty = next(iter(nn_module_stack.values()))
+                assert issubclass(ty, torch.nn.Module)
+                assert path != root
+                assert root in path
+            # TODO(zhxchen17) normalize all the way down to dot strings.
+            node.meta["nn_module_stack"] = {'L__self__': (root, root_cls), **nn_module_stack}
```
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:240]: NIT: Can you leave a comment on what you are changing in the spec?

File: torch/_export/__init__.py [hunk 4]
```diff
@@ -352,6 +382,26 @@ def export(
...
+        param_lookup = {}
+        buffer_lookup = {}
+        for name, param in f.named_parameters():
+            param_lookup[id(param)] = name
+        for name, buffer in f.named_buffers():
+            buffer_lookup[id(buffer)] = name
+        for dynamo_name, dynamo_param in gm_torch_level.named_parameters():
+            assert dynamo_name not in param_buffer_table
+            if id(dynamo_param) in param_lookup:
+                param_buffer_table[dynamo_name] = param_lookup[id(dynamo_param)]
```
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:396]: Are you mapping dynamo param name to the original name?

File: torch/_export/__init__.py [hunk 5]
```diff
@@ -210,6 +210,36 @@ def _safe_to_skip_dynamo(gm: torch.fx.GraphModule):
     return False
 
 
+def _replace_param_buffer_names(param_buffer_table, sig):
+    replace = lambda x: param_buffer_table.get(x, x)
+    sig.parameters = pytree.tree_map(replace, sig.parameters)
```
Comment (tugsbayasgalan) [inline torch/_export/__init__.py:215]: OP way to replace :)

File: torch/_export/__init__.py (hunk 1 above)
Comment (zhxchen17) [reply torch/_export/__init__.py:513]: I think range_constraints is for both unbacked and backed symints?

**Review by tugsbayasgalan — APPROVED** [review]

File: torch/_export/__init__.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_export/__init__.py:513]: Yeah thanks for confirming!

File: torch/_export/__init__.py (hunk 4 above)
Comment (zhxchen17) [reply torch/_export/__init__.py:396]: yes

**Comment by desertfire** [comment]
@zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

**Comment by zhxchen17** [comment]
> @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

Can I take a look and see if this forward fixable?

**Comment by zhxchen17** [comment]
> @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?

@desertfire yeah I didn't see oss test failures.

**Comment by desertfire** [comment]
> > @zhxchen17 , `python benchmarks/dynamo/huggingface.py --bfloat16 --accuracy --inference --device cuda --export --only  BertForMaskedLM` (and so are other benchmarks) fails after this PR. Reverting this PRK will make it pass. I guess we don't have any CI tasks for testing `--export` and `--export-aot-inductor` on the OSS benchmarks?
> 
> Can I take a look and see if this forward fixable?

Go ahead, since it doesn't affect any CI at the moment, but if it takes more than a few hours, let's revert. And I will add CI tests after your fix or revert.

**Comment by zhxchen17** [comment]
@desertfire looks like there's a small bug, I'm fixing it in https://github.com/pytorch/pytorch/pull/108354.
